src/beatree/leaf/store.rs

```diff
@@ -0,0 +1,225 @@
+    pub fn start_tx<'a>(&'a mut self) -> LeafStoreTx {
+        LeafStoreTx {
+            nonce_page_number: &mut self.nonce_page_number,
+            max_page_number: self.max_page_number,
+            free_list: &mut self.free_list,
+            to_allocate: vec![],
+            released: vec![],
+            exceeded: vec![],
+        }
+    }
+
+    // updates the mex supported page number based on the leaf store file size
+    pub fn update_size(&mut self) {
+        let file_size = self
+            .store_file
+            .metadata()
+            .expect("Error extracting metadata from LeafStore file")
+            .size() as usize;
+
+        self.max_page_number = PageNumber((file_size / PAGE_SIZE) as u32);
+    }
+
+    // perform writes for all provided pages, then waits till the end
+    //
+    // panics if the provided PageNumber is bigger then the max supported one
+    pub fn write(&mut self, pages: Vec<(PageNumber, Box<Page>)>) {
+        let n_write_requests = pages.len();
+        for (pn, page) in pages {
+            let mut command = Some(IoCommand {
+                kind: IoKind::Write(self.store_file.as_raw_fd(), pn.0 as u64, page),
+                handle: self.io_handle_index,
+                user_data: 0,
+            });
+
+            while let Some(c) = command.take() {
+                match self.io_sender.try_send(c) {
+                    Ok(()) => break,
+                    Err(TrySendError::Disconnected(_)) => panic!("I/O leaf store worker dropped"),
+                    Err(TrySendError::Full(c)) => {
+                        command = Some(c);
+                    }
+                }
+            }
+        }
+
+        // wait for completion of all write commands
+        for _ in 0..n_write_requests {
+            let completion = self
+                .io_receiver
+                .recv()
+                .expect("I/O leaf store worker dropped");
+            assert!(completion.result.is_ok());
+        }
+    }
+}
+
+pub struct LeafStoreTx<'a> {
+    nonce_page_number: &'a mut PageNumber,
+    max_page_number: PageNumber,
+    free_list: &'a mut FreeList,
+    released: Vec<PageNumber>,
+    to_allocate: Vec<(PageNumber, LeafPage)>,
+    exceeded: Vec<(PageNumber, LeafPage)>,
+}
+
+struct LeafStoreTxOutput {
+    // contains pages that can already be written in the LeafStore
+    to_allocate: Vec<(PageNumber, Box<Page>)>,
+    // contains pages that require the LeafStore to increase
+    exceeded: Vec<(PageNumber, Box<Page>)>,
+    new_free_list_head: PageNumber,
+}
+
+impl<'a> LeafStoreTx<'a> {
+    pub fn allocate(&mut self, leaf_page: LeafPage) -> PageNumber {
+        let leaf_pn = match self.free_list.pop() {
+            Some(pn) => pn,
+            None => {
+                let pn = *self.nonce_page_number;
+                self.nonce_page_number.0 += 1;
+                pn
+            }
+        };
+
+        if leaf_pn.0 < self.max_page_number.0 {
```

Comment:
Could you explain/leave a comment why is it `<` in contrast to `≤`?

Reply:
that's just because I did 

```rust
        self.max_page_number = PageNumber((file_size / PAGE_SIZE) as u32);
       // instead of
        self.max_page_number = PageNumber((file_size / PAGE_SIZE) as u32 - 1);
```

the second option require a check of the file size (it must be at least bigger then the page_size) but then it would require `<=` when comparing it with `leaf_pn`

I will update anyway the code to use `<=`

Reply:
I changed my mind, now I remember why I decided to use just `<`, because in the case of the file being empty, a `max_page_number` equal to 0 will put every create page into the `exceeded` field, while if it would be `<=` the `max_page_number` should become something like an option to handle the empty file

Reply:
yeah would be useful to leave a comment about that.